File: GroupPlan.swift
//
//  GroupPlan.swift
//  Avenue3
//

import Foundation

/// An invitee's response to a group plan invite.
enum GroupPlanResponse: String, Codable {
    case invited
    case going
    case cantMake
}

/// Lifecycle status of a group plan.
enum GroupPlanStatus: String, Codable {
    case active
    case cancelled
}

/// A hangout plan proposed by one host to several matches at once. Distinct from `Plan`
/// (1-on-1 proposals): no counter-proposal flow, no messaging thread — the host picks a
/// single date up front and invitees independently say Going or Can't make it.
struct GroupPlan: Identifiable, Codable {
    let id: String

    /// User who created the invite.
    let hostID: String

    /// Every user invited to this plan.
    let inviteeIDs: [String]

    /// Each invitee's response, keyed by user ID. Every invitee starts as `.invited`.
    var responses: [String: GroupPlanResponse]

    /// Selected activity for the hangout.
    let activity: Activity

    /// Optional location (free text) — same meaning as on `Plan`.
    let location: String?

    /// Name of the place selected via location search (nil for free-typed locations).
    let locationName: String?

    /// Coordinates of the place selected via location search, for tap-to-navigate.
    let locationLatitude: Double?
    let locationLongitude: Double?

    /// The single date/time for this plan.
    let date: Date

    /// Status of the plan.
    var status: GroupPlanStatus

    let createdAt: Date
    var updatedAt: Date

    init(
        id: String = UUID().uuidString,
        hostID: String,
        inviteeIDs: [String],
        responses: [String: GroupPlanResponse]? = nil,
        activity: Activity,
        location: String? = nil,
        locationName: String? = nil,
        locationLatitude: Double? = nil,
        locationLongitude: Double? = nil,
        date: Date,
        status: GroupPlanStatus = .active,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.hostID = hostID
        self.inviteeIDs = inviteeIDs
        self.responses = responses ?? Dictionary(uniqueKeysWithValues: inviteeIDs.map { ($0, .invited) })
        self.activity = activity
        self.location = location
        self.locationName = locationName
        self.locationLatitude = locationLatitude
        self.locationLongitude = locationLongitude
        self.date = date
        self.status = status
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    func isHost(userID: String) -> Bool { hostID == userID }
    func isInvitee(userID: String) -> Bool { inviteeIDs.contains(userID) }
    func response(for userID: String) -> GroupPlanResponse? { responses[userID] }

    var goingCount: Int { responses.values.filter { $0 == .going }.count }
}

File: android/app/src/main/java/com/georgeappdev/atxfriends/data/model/GroupPlan.kt
package com.georgeappdev.atxfriends.data.model

import com.georgeappdev.atxfriends.data.firestore.DocReader
import com.georgeappdev.atxfriends.data.firestore.GroupPlanFields
import java.time.Instant

/** `groupPlans/{planID}` — the multi-invitee plan shown on the Upcoming tab. */
data class GroupPlan(
    val id: String,
    val hostID: String,
    val inviteeIDs: List<String>,
    /** Keyed by invitee user ID. */
    val responses: Map<String, GroupPlanResponse>,
    val activity: Activity,
    val location: String?,
    val locationName: String?,
    val locationLatitude: Double?,
    val locationLongitude: Double?,
    val date: Instant,
    val status: GroupPlanStatus,
    val createdAt: Instant,
    val updatedAt: Instant,
) {
    companion object {
        /**
         * Same required fields as iOS `GroupPlansService`, except unrecognized `status` or
         * response strings decode to UNKNOWN instead of being dropped.
         */
        fun fromFirestore(id: String, data: Map<String, Any?>): GroupPlan? {
            val r = DocReader(data)
            val responses = r.map(GroupPlanFields.RESPONSES) ?: return null
            return GroupPlan(
                id = id,
                hostID = r.string(GroupPlanFields.HOST_ID) ?: return null,
                inviteeIDs = r.stringList(GroupPlanFields.INVITEE_IDS) ?: return null,
                responses = responses.mapValues { (_, v) -> GroupPlanResponse.fromRaw(v as? String) },
                activity = Activity.fromFirestore(r.map(GroupPlanFields.ACTIVITY)) ?: return null,
                date = r.instant(GroupPlanFields.DATE) ?: return null,
                status = GroupPlanStatus.fromRaw(r.string(GroupPlanFields.STATUS) ?: return null),
                createdAt = r.instant(GroupPlanFields.CREATED_AT) ?: return null,
                updatedAt = r.instant(GroupPlanFields.UPDATED_AT) ?: return null,
                location = r.string(GroupPlanFields.LOCATION),
                locationName = r.string(GroupPlanFields.LOCATION_NAME),
                locationLatitude = r.double(GroupPlanFields.LOCATION_LATITUDE),
                locationLongitude = r.double(GroupPlanFields.LOCATION_LONGITUDE),
            )
        }
    }
}
